discovery/premature_cache.go

```diff
@@ -0,0 +1,266 @@
+}
+
+// prematureCache defines an interface that handles reading and writing
+// premature network messages to the relaying node's cache.
+type prematureCache[T messageKey] interface {
+	// fetchMessage reads and removes the cached messages from the relaying
+	// node's cache. It takes a public key to query the node's cache and
+	// reads the value using the key.
+	fetchMessage(cacheKey [33]byte, key T) (*cachedNetworkMsg, error)
+
+	// writeMessage writes a message to the relaying node's cache. It will
+	// fetch the node's cache using the public key of the message's sender,
+	// and writes the message based on the `lnwire.Message` type.
+	writeMessage(nMsg *networkMsg) error
+}
+
+// cachedNetworkMsg is a wrapper around a network message that can be used with
+// *lru.Cache.
+type cachedNetworkMsg struct {
+	msgs []*processedNetworkMsg
+}
+
+// Size returns the "size" of an entry. We return the number of items as we
+// just want to limit the total amount of entries rather than do accurate size
+// accounting.
+//
+// NOTE: part of the cache.Value interface.
+func (c *cachedNetworkMsg) Size() (uint64, error) {
+	return uint64(len(c.msgs)), nil
+}
+
+// relayingNodeCache implements `prematureCache` interface. It is responsible
+// for managing premature network messages by each sender.
+type relayingNodeCache[T messageKey] struct {
+	sync.Mutex
+
+	// cacheMap keeps a track of caches by the relaying node's public key.
+	cacheMap map[[33]byte]*lru.Cache
+}
+
+// newChannelUpdateCache creates a new `relayingNodeCache` using short channel
+// ID.
+func newChannelUpdateCache() *relayingNodeCache[uint64] {
+	return &relayingNodeCache[uint64]{
+		cacheMap: make(map[[33]byte]*lru.Cache),
+	}
+}
+
+// newNodeAnnouncementCache creates a new `relayingNodeCache` using public key
+// bytes.
+func newNodeAnnouncementCache() *relayingNodeCache[[33]byte] {
+	return &relayingNodeCache[[33]byte]{
+		cacheMap: make(map[[33]byte]*lru.Cache),
+	}
+}
+
+// getOrCreateCache initializes a new cache for the given node, or returns the
+// cache if it's already been initialized.
+func (r *relayingNodeCache[T]) getOrCreateCache(pub [33]byte) *lru.Cache {
+	r.Lock()
+	defer r.Unlock()
+
+	// Get the node cache.
+	nc, ok := r.cacheMap[pub]
+	if !ok {
+		// If the cache is not created, initialize one and put it in
+		// the map.
+		nc = lru.NewCache(maxPrematureUpdates)
+		r.cacheMap[pub] = nc
+
+		log.Debugf("Created premature cache for relaying node %x", pub)
+	}
+
+	return nc
+}
+
+// fetchMessage queries the relaying node's cache using the key and returns the
+// messages found. The messages are removed once read from the cache. Returns
+// an empty `cachedNetworkMsg` if nothing is found.
+func (r *relayingNodeCache[T]) fetchMessage(pub [33]byte,
+	key T) (*cachedNetworkMsg, error) {
+
+	// Read the messages.
+	cachedMsg, err := r.readMessage(pub, key)
+	if err != nil {
+		return nil, err
+	}
+
+	// Exit early if nothing is found.
+	if cachedMsg.msgs == nil {
```

Comment:
This'll panic if `readMessageFromCache` returns an error, which means this gets a nil value after `readMessage`.

Reply:
Then it won't get here as the `r.readMessage` would exit on the error.